copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+			}
+			// chase links. if we hit a dead end, we should just fail
+			followedLinks := 0
+			const maxFollowedLinks = 16
+			for info.Mode()&os.ModeType == os.ModeSymlink && followedLinks < maxFollowedLinks {
+				path, err := os.Readlink(item)
+				if err != nil {
+					continue
+				}
+				if filepath.IsAbs(path) {
+					path = filepath.Join(request.Root, path)
+				} else {
+					path = filepath.Join(filepath.Dir(item), path)
+				}
+				item = path
+				if _, err = filepath.Rel(request.Root, item); err != nil {
+					return errors.Wrapf(err, "copier: get: computing path of %q(%q) relative to %q", queue[i], item, request.Root)
+				}
+				if info, err = os.Lstat(item); err != nil {
+					return errors.Wrapf(err, "copier: get: lstat %q(%q)", queue[i], item)
+				}
+				followedLinks++
+			}
+			if followedLinks >= maxFollowedLinks {
+				return errors.Wrapf(syscall.ELOOP, "copier: get: resolving symlink %q(%q)", queue[i], item)
+			}
+			// evaluate excludes relative to the root directory
+			if info.Mode().IsDir() {
+				walkfn := func(path string, info os.FileInfo, err error) error {
+					// compute the path of this item
+					// relative to the top-level directory,
+					// for the tar header
+					rel, relErr := filepath.Rel(item, path)
+					if relErr != nil {
+						return errors.Wrapf(relErr, "copier: get: error computing path of %q relative to top directory %q", path, item)
+					}
+					if err != nil {
+						return errors.Wrapf(err, "copier: get: error reading %q", path)
+					}
+					// prefix the original item's name if we're keeping it
+					if relNamePrefix != "" {
+						rel = filepath.Join(relNamePrefix, rel)
+					}
+					if rel == "" || rel == "." {
+						// skip the "." entry
+						return nil
+					}
+					_, skip, err := pathIsExcluded(request.Root, path, pm)
+					if err != nil {
+						return err
+					}
+					if skip {
+						// don't use filepath.SkipDir
+						// here, since a more specific
+						// but-include-this for
+						// something under it might
+						// also be in the excludes list
+						return nil
+					}
+					// if it's a symlink, read its target
+					symlinkTarget := ""
+					if info.Mode()&os.ModeType == os.ModeSymlink {
+						target, err := os.Readlink(path)
+						if err != nil {
+							return errors.Wrapf(err, "copier: get: readlink(%q(%q))", rel, path)
+						}
+						symlinkTarget = target
+					}
+					// add the item to the outgoing tar stream
+					return copierHandlerGetOne(info, symlinkTarget, rel, path, request.GetOptions, tw, hardlinkChecker, idMappings)
+				}
+				// walk the directory tree, checking/adding items individually
+				if err := filepath.Walk(item, walkfn); err != nil {
+					return errors.Wrapf(err, "copier: get: %q(%q)", queue[i], item)
+				}
+				itemsCopied++
+			} else {
+				_, skip, err := pathIsExcluded(request.Root, item, pm)
+				if err != nil {
+					return err
+				}
+				if skip {
+					return nil
```

Comment:
This completely ignores the rest of `queue`.

Reply:
Not sure I follow.  This is the callback that `filepath.Walk()` calls, so returning `nil` looks like a normal result.

Reply:
Do you mean the callback that starts at line 917:
```go
			if info.Mode().IsDir() {
 				walkfn := func(path string, info os.FileInfo, err error) error {
```
?

At least based on indentation, that ends on line 959:
```go
					return copierHandlerGetOne(info, symlinkTarget, rel, path, request.GetOptions, tw, hardlinkChecker, idMappings)
 				}
 				// walk the directory tree, checking/adding items individually
```

Reply:
Eek, you're right.